File: 2019/03/03.clj
(ns dec03)
  (require '[clojure.string :as string])
  (require '[clojure.set :as sets])

(defn parse-int [n] (Integer/parseInt (str n)))

(defn parse-file [file]
  (->> (slurp file)
       (string/split-lines)
       (map #(string/split % #","))))

(defn direction->deltas [d]
  (case d
    \R [1 0]
    \L [-1 0]
    \U [0 1]
    \D [0 -1]))

(defn parse-direction [d]
  [(direction->deltas (first d))
   (parse-int (subs d 1))])

(defn add-coordinates [a b]
  [(+ (first a) (first b) )
   (+ (second a) (second b))])

(defn distance [[x y]] (+ (Math/abs x) (Math/abs y)))

(defn step->coordinates [coordinates [delta steps]]
  (loop [n 0
         coords coordinates
         current (add-coordinates (last coordinates) delta)]
    (if (= n steps)
      coords
      (recur (inc n)
             (conj coords current)
             (add-coordinates current delta)))))

(defn path->coordinates [path]
  (reduce step->coordinates [[0 0]] path))

(defn row->coords [r]
  (->> (map parse-direction r)
       (path->coordinates)
       (rest)))

(defn position-of [coordinate]
    (fn [wire] (.indexOf wire coordinate)))

(defn distance-along [wires coordinate]
    (reduce + 
        (map (position-of coordinate) wires)))

(def wires 
    (->> (parse-file "input.path")
         (map row->coords)))

(def intersections 
    (reduce sets/intersection (map set wires)))

(println "Part 1:" 
    (->> (map distance intersections)
         (set)
         (#(disj % 0))
         (reduce min)))

(println "Part 2:"
    (reduce 
        (fn [acc point]
            (min acc (distance-along wires point)))
        ##Inf
        intersections))

File: 2015/03/03.py
with open("input.txt") as f:
    path = f.readline()

deltas = {
    '^': (0, 1),
    '>': (1, 0),
    'v': (0, -1),
    '<': (-1, 0),
}

s_x, s_y = 0, 0
r_x, r_y = 0, 0
x, y = 0, 0
first_visited = {(0, 0)}
second_visited = {(0, 0)}
for i, c in enumerate(path):
    dx, dy = deltas[c]
    x, y = x + dx, y + dy
    first_visited.add((x, y))

    if i % 2 == 0:
        dx, dy = deltas[c]
        s_x, s_y = s_x + dx, s_y + dy
        second_visited.add((s_x, s_y))
    else:
        dx, dy = deltas[c]
        r_x, r_y = r_x + dx, r_y + dy
        second_visited.add((r_x, r_y))

print("Part 1:", len(first_visited))
print("Part 2:", len(second_visited))